{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0cd66d6c",
   "metadata": {},
   "source": [
    "# 生成华大版本的bgef"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4154bb7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting gem2gef_bgi_bug.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile gem2gef_bgi_bug.py\n",
    "from gefpy import bgef_writer_cy\n",
    "bgef_writer_cy.generate_bgef(\n",
    "    input_file = \"out/Y00855N1.tissue.gem.gz\",\n",
    "    bgef_file = \"out/Y00855N1_bgi.bgef\",\n",
    "    bin_sizes = [1],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00c5e584",
   "metadata": {},
   "source": [
    "上述代码会报错\n",
    "> (st) PS D:\\Rust\\geftools_rs> & E:/Anaconda_envs/envs/st/python.exe d:/Rust/geftools_rs/gem2gef_bgi_bug.py \\\n",
    "geneID  x       y       MIDCount        ExonCount\\\n",
    " 5\\\n",
    "create bgef file: out/Y00855N1_bgi.bgef\\\n",
    "can not find sn info from input file out/Y00855N1.tissue.gem.gz\\\n",
    "bin 1 matrix: min_x=0 len_x=-2147483648 min_y=0 len_y=-2147483648 matrix_len=0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ef69c6c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting gem2gef_bgi_bug.py\n"
     ]
    }
   ],
   "source": [
    "%%writefile gem2gef_bgi_bug.py\n",
    "from gefpy import bgef_writer_cy\n",
    "bgef_writer_cy.generate_bgef(\n",
    "    input_file = \"out/test10000.gem.gz\",\n",
    "    bgef_file = \"out/test10000_bgi.bgef\",\n",
    "    bin_sizes = [1],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "09370afc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'d:\\\\Rust\\\\geftools_rs'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Group] geneExp\n",
      "[Group] geneExp/bin1\n",
      "[Dataset] geneExp/bin1/exon  shape=(8325,)  dtype=uint16\n",
      "[Dataset] geneExp/bin1/expression  shape=(8325,)  dtype=[('x', '<i4'), ('y', '<i4'), ('count', '<u2')]\n",
      "[Dataset] geneExp/bin1/gene  shape=(3051,)  dtype=[('geneID', 'S64'), ('geneName', 'S64'), ('offset', '<u4'), ('count', '<u4')]\n",
      "[Group] stat\n",
      "[Dataset] stat/gene  shape=(3051,)  dtype=[('geneID', 'S64'), ('geneName', 'S64'), ('MIDcount', '<u4'), ('E10', '<f4')]\n",
      "[Group] wholeExp\n",
      "[Dataset] wholeExp/bin1  shape=(21270, 21263)  dtype=[('MIDcount', '<u4'), ('genecount', '<u2')]\n",
      "[Group] wholeExpExon\n",
      "[Dataset] wholeExpExon/bin1  shape=(21270, 21263)  dtype=uint32\n",
      "\n",
      "Root attributes:\n",
      "  bin_type: [b'Bin'] (type=ndarray)\n",
      "  gef_area: [0.] (type=ndarray)\n",
      "  geftool_ver: [ 1  1 20] (type=ndarray)\n",
      "  omics: [b'Transcriptomics'] (type=ndarray)\n",
      "  version: [4] (type=ndarray)\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import h5py\n",
    "import numpy as np\n",
    "display(os.getcwd())\n",
    "\n",
    "def print_structure(name, obj):\n",
    "    if isinstance(obj, h5py.Group):\n",
    "        print(f\"[Group] {name}\")\n",
    "    elif isinstance(obj, h5py.Dataset):\n",
    "        print(f\"[Dataset] {name}  shape={obj.shape}  dtype={obj.dtype}\")\n",
    "\n",
    "path = \"out/test10000_bgi.bgef\"\n",
    "with h5py.File(path, \"r\") as f:\n",
    "    f.visititems(print_structure)\n",
    "\n",
    "    print(\"\\nRoot attributes:\")\n",
    "    for k, v in f.attrs.items():\n",
    "        # bytes → 字符串\n",
    "        if isinstance(v, bytes):\n",
    "            v = v.decode()\n",
    "        # numpy 标量 → 提取值但保留 dtype\n",
    "        if isinstance(v, np.generic):\n",
    "            print(f\"  {k}: {v.item()} (dtype={v.dtype})\")\n",
    "        else:\n",
    "            print(f\"  {k}: {v} (type={type(v).__name__})\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "dc435676",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>>Attribute:\n",
      "lenX: 17456, int32\n",
      "lenY: 18054, int32\n",
      "maxExon: 25, uint32\n",
      "maxGene: 47, uint32\n",
      "maxMID: 49, uint32\n",
      "minX: 3767, int32\n",
      "minY: 3330, int32\n",
      "number: 85019725, uint64\n",
      "resolution: 500, uint32\n",
      ">>>Dataset:\n",
      "(17456, 18054) [[(0, 0) (0, 0) (0, 0) ... (0, 0) (0, 0) (0, 0)]\n",
      " [(0, 0) (0, 0) (0, 0) ... (0, 0) (0, 0) (0, 0)]\n",
      " [(0, 0) (0, 0) (0, 0) ... (0, 0) (0, 0) (0, 0)]\n",
      " ...\n",
      " [(0, 0) (0, 0) (0, 0) ... (0, 0) (0, 0) (0, 0)]\n",
      " [(1, 1) (2, 2) (0, 0) ... (3, 3) (4, 3) (1, 1)]\n",
      " [(0, 0) (0, 0) (0, 0) ... (0, 0) (1, 1) (0, 0)]]\n"
     ]
    }
   ],
   "source": [
    "# 代码示例\n",
    "import h5py\n",
    "import numpy as np\n",
    "\n",
    "path = \"out/Y00855N1.bgef\"\n",
    "with h5py.File(path, \"r\") as f:\n",
    "    h5path = \"wholeExp/bin1\"\n",
    "\n",
    "    # 输出属性\n",
    "    print(\">>>Attribute:\")\n",
    "    for k, v in f[h5path].attrs.items():\n",
    "        # 一些属性是 bytes，需要手动 decode\n",
    "        if isinstance(v, bytes):\n",
    "            v = v.decode()\n",
    "        print(f\"{k}: {v}, {v.dtype}\")\n",
    "    \n",
    "    # 数据矩阵\n",
    "    print(\">>>Dataset:\")\n",
    "    expr = np.array(f[h5path]) # 替换这里f的键值即可\n",
    "    print(expr.shape, expr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c764c19",
   "metadata": {},
   "source": [
    "# 查看输出hdf5数据结构"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "25d8f05f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'d:\\\\Rust\\\\geftools_rs'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Group] geneExp\n",
      "[Group] geneExp/bin1\n",
      "[Dataset] geneExp/bin1/exon  shape=(153549346,)  dtype=uint32\n",
      "[Dataset] geneExp/bin1/expression  shape=(153549346,)  dtype=[('x', '<i4'), ('y', '<i4'), ('count', '<u4')]\n",
      "[Dataset] geneExp/bin1/gene  shape=(40417,)  dtype=[('geneID', 'S64'), ('geneName', 'S64'), ('offset', '<u4'), ('count', '<u4')]\n",
      "[Group] wholeExp\n",
      "[Dataset] wholeExp/bin1  shape=(17456, 18054)  dtype={'names': ['MIDcount', 'genecount'], 'formats': ['<u4', '<u2'], 'offsets': [0, 4], 'itemsize': 8}\n",
      "[Group] wholeExpExon\n",
      "[Dataset] wholeExpExon/bin1  shape=(17456, 18054)  dtype=uint32\n",
      "\n",
      "Root attributes:\n",
      "  bin_type: Bin (type=str)\n",
      "  gef_area: 0.0 (dtype=float32)\n",
      "  geftool_ver: [ 1  1 20] (type=ndarray)\n",
      "  omics: Transcriptomics (type=str)\n",
      "  sn: Y00855N1 (type=str)\n",
      "  version: 4 (dtype=uint32)\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import h5py\n",
    "display(os.getcwd())\n",
    "\n",
    "def print_structure(name, obj):\n",
    "    if isinstance(obj, h5py.Group):\n",
    "        print(f\"[Group] {name}\")\n",
    "    elif isinstance(obj, h5py.Dataset):\n",
    "        print(f\"[Dataset] {name}  shape={obj.shape}  dtype={obj.dtype}\")\n",
    "\n",
    "path = \"out/Y00855N1.bgef\"\n",
    "with h5py.File(path, \"r\") as f:\n",
    "    f.visititems(print_structure)\n",
    "\n",
    "    print(\"\\nRoot attributes:\")\n",
    "    for k, v in f.attrs.items():\n",
    "        # bytes → 字符串\n",
    "        if isinstance(v, bytes):\n",
    "            v = v.decode()\n",
    "        # numpy 标量 → 提取值但保留 dtype\n",
    "        if isinstance(v, np.generic):\n",
    "            print(f\"  {k}: {v.item()} (dtype={v.dtype})\")\n",
    "        else:\n",
    "            print(f\"  {k}: {v} (type={type(v).__name__})\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8873769f",
   "metadata": {},
   "source": [
    "Y00855ND.tissue.gem.gz文件总行数是123196419"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f177f8dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>>Attribute:\n",
      "lenX: 17456, int32\n",
      "lenY: 18054, int32\n",
      "maxExon: 25, uint32\n",
      "maxGene: 47, uint32\n",
      "maxMID: 49, uint32\n",
      "minX: 3767, int32\n",
      "minY: 3330, int32\n",
      "number: 85019725, uint64\n",
      "resolution: 500, uint32\n",
      ">>>Dataset:\n",
      "(17456, 18054) [[(0, 0) (0, 0) (0, 0) ... (0, 0) (0, 0) (0, 0)]\n",
      " [(0, 0) (0, 0) (0, 0) ... (0, 0) (0, 0) (0, 0)]\n",
      " [(0, 0) (0, 0) (0, 0) ... (0, 0) (0, 0) (0, 0)]\n",
      " ...\n",
      " [(0, 0) (0, 0) (0, 0) ... (0, 0) (0, 0) (0, 0)]\n",
      " [(1, 1) (2, 2) (0, 0) ... (3, 3) (4, 3) (1, 1)]\n",
      " [(0, 0) (0, 0) (0, 0) ... (0, 0) (1, 1) (0, 0)]]\n"
     ]
    }
   ],
   "source": [
    "# 代码示例\n",
    "import h5py\n",
    "import numpy as np\n",
    "path = \"out/test.bgef\"\n",
    "path = \"out/Y00855N1.bgef\"\n",
    "with h5py.File(path, \"r\") as f:\n",
    "    # h5path = \"/geneExp/bin1/expression\"\n",
    "    h5path = \"wholeExp/bin1\"\n",
    "\n",
    "    # 输出属性\n",
    "    print(\">>>Attribute:\")\n",
    "    for k, v in f[h5path].attrs.items():\n",
    "        # 一些属性是 bytes，需要手动 decode\n",
    "        if isinstance(v, bytes):\n",
    "            v = v.decode()\n",
    "        print(f\"{k}: {v}, {v.dtype}\")\n",
    "    \n",
    "    # 数据矩阵\n",
    "    print(\">>>Dataset:\")\n",
    "    expr = np.array(f[h5path]) # 替换这里f的键值即可\n",
    "    print(expr.shape, expr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38787eb8",
   "metadata": {},
   "source": [
    "# 查看cgef结构"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4aee1240",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'d:\\\\Rust\\\\geftools_rs'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Group] cellBin\n",
      "[Dataset] cellBin/blockIndex  shape=(4900,)  dtype=uint32\n",
      "[Dataset] cellBin/blockSize  shape=(4,)  dtype=uint32\n",
      "[Dataset] cellBin/cell  shape=(98651,)  dtype=[('id', '<u4'), ('x', '<i4'), ('y', '<i4'), ('offset', '<u4'), ('geneCount', '<u2'), ('expCount', '<u2'), ('dnbCount', '<u2'), ('area', '<u2'), ('cellTypeID', '<u2'), ('clusterID', '<u2')]\n",
      "[Dataset] cellBin/cellBorder  shape=(98651, 32, 2)  dtype=int16\n",
      "[Dataset] cellBin/cellExon  shape=(98651,)  dtype=uint16\n",
      "[Dataset] cellBin/cellExp  shape=(18658002,)  dtype=[('geneID', '<u4'), ('count', '<u2')]\n",
      "[Dataset] cellBin/cellExpExon  shape=(18658002,)  dtype=uint16\n",
      "[Dataset] cellBin/cellTypeList  shape=(1,)  dtype=|S32\n",
      "[Dataset] cellBin/gene  shape=(40417,)  dtype=[('geneID', 'S64'), ('geneName', 'S64'), ('offset', '<u4'), ('cellCount', '<u4'), ('expCount', '<u4'), ('maxMIDcount', '<u2')]\n",
      "[Dataset] cellBin/geneExon  shape=(40417,)  dtype=uint32\n",
      "[Dataset] cellBin/geneExp  shape=(18658002,)  dtype=[('cellID', '<u4'), ('count', '<u2')]\n",
      "[Dataset] cellBin/geneExpExon  shape=(18658002,)  dtype=uint16\n",
      "\n",
      "Root attributes:\n",
      "  bin_type: [b'CellBin'] (type=ndarray)\n",
      "  geftool_ver: [ 1  1 20] (type=ndarray)\n",
      "  maxX: 21222 (dtype=int32)\n",
      "  maxY: 21383 (dtype=int32)\n",
      "  offsetX: [3767] (type=ndarray)\n",
      "  offsetY: [3330] (type=ndarray)\n",
      "  omics: [b''] (type=ndarray)\n",
      "  resolution: [500] (type=ndarray)\n",
      "  sn: Y00855N1 (type=str)\n",
      "  version: [4] (type=ndarray)\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import h5py\n",
    "display(os.getcwd())\n",
    "\n",
    "def print_structure(name, obj):\n",
    "    if isinstance(obj, h5py.Group):\n",
    "        print(f\"[Group] {name}\")\n",
    "    elif isinstance(obj, h5py.Dataset):\n",
    "        print(f\"[Dataset] {name}  shape={obj.shape}  dtype={obj.dtype}\")\n",
    "\n",
    "path = \"out/cell_cut_out/Y00855N1.cellbin.gef\"\n",
    "with h5py.File(path, \"r\") as f:\n",
    "    f.visititems(print_structure)\n",
    "\n",
    "    print(\"\\nRoot attributes:\")\n",
    "    for k, v in f.attrs.items():\n",
    "        # bytes → 字符串\n",
    "        if isinstance(v, bytes):\n",
    "            v = v.decode()\n",
    "        # numpy 标量 → 提取值但保留 dtype\n",
    "        if isinstance(v, np.generic):\n",
    "            print(f\"  {k}: {v.item()} (dtype={v.dtype})\")\n",
    "        else:\n",
    "            print(f\"  {k}: {v} (type={type(v).__name__})\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8b780b09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">>>Attribute:\n",
      "maxCellCount: [77376], uint32\n",
      "maxExpCount: [65272], uint32\n",
      "minCellCount: [1], uint32\n",
      "minExpCount: [1], uint32\n",
      ">>>Dataset:\n",
      "(40417,) [(b'0610005C13Rik', b'0610005C13Rik',        0,  61,  66, 2)\n",
      " (b'0610006L08Rik', b'0610006L08Rik',       61,  38,  38, 1)\n",
      " (b'0610009B22Rik', b'0610009B22Rik',       99, 479, 513, 3) ...\n",
      " (b'n-R5s97', b'n-R5s97',        0,   0,   0, 0)\n",
      " (b'n-R5s98', b'n-R5s98', 18658001,   1,   1, 1)\n",
      " (b'n-TSaga9', b'n-TSaga9',        0,   0,   0, 0)]\n"
     ]
    }
   ],
   "source": [
    "# 代码示例\n",
    "import h5py\n",
    "import numpy as np\n",
    "\n",
    "path = \"out/cell_cut_out/Y00855N1.cellbin.gef\"\n",
    "with h5py.File(path, \"r\") as f:\n",
    "    h5path = \"cellBin/gene\"\n",
    "\n",
    "    # 输出属性\n",
    "    print(\">>>Attribute:\")\n",
    "    for k, v in f[h5path].attrs.items():\n",
    "        # 一些属性是 bytes，需要手动 decode\n",
    "        if isinstance(v, bytes):\n",
    "            v = v.decode()\n",
    "        print(f\"{k}: {v}, {v.dtype}\")\n",
    "    \n",
    "    # 数据矩阵\n",
    "    print(\">>>Dataset:\")\n",
    "    expr = np.array(f[h5path]) # 替换这里f的键值即可\n",
    "    print(expr.shape, expr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8108f8fc",
   "metadata": {},
   "source": [
    "# 使用stereopy读取和显示"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a9824bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import pandas as pd\n",
    "from natsort import natsorted\n",
    "import stereo as st\n",
    "from stereo.core.ms_data import MSData\n",
    "from stereo.core.ms_pipeline import slice_generator\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "targets = pd.read_csv(\"data/targets.csv\")\n",
    "# construct MSData object\n",
    "ms_data = MSData(_relationship='other', _var_type='intersect')\n",
    "\n",
    "for idx, row in targets.iterrows():\n",
    "    print(row['acc_id'])\n",
    "    # read the gem file\n",
    "    data_path = f\"data/cell_cut_out/{row['acc_id']}.cellbin.gef\"\n",
    "    # 原始GEM没有cellbin这一项，所以只能按bins读取\n",
    "    ms_data += st.io.read_gef(\n",
    "        file_path=data_path,\n",
    "        bin_type='cell_bins',\n",
    "        is_sparse=True,\n",
    "        # gene_name_index=True, # 手搓的bgef没有真正的geneID\n",
    "        )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "st",
   "language": "python",
   "name": "st"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
